src/libsyntax/parse/lexer/tokentrees.rs

```diff
@@ -49,9 +49,23 @@ impl<'a> StringReader<'a> {
                 let msg = "this file contains an un-closed delimiter";
                 let mut err = self.sess.span_diagnostic.struct_span_err(self.span, msg);
                 for &(_, sp) in &self.open_braces {
-                    err.span_help(sp, "did you mean to close this delimiter?");
+                    err.span_label(sp, "un-closed delimiter");
                 }
 
+                if let Some((delim, _)) = self.open_braces.last() {
+                    if let Some((d, open_sp, close_sp)) = self.suspicious_open_spans.iter()
+                        .filter(|(d, _, _)| delim == d)
+                        .next()  // these are in reverse order as they get inserted on close, but
+                    {            // we want the last open/first close
+                        if d == delim {
+                            err.span_label(*open_sp, "this might be the culprit...");
```

Comment:
I love this language, but perhaps something a touch more formal...

- "this delimiter might not be properly closed" ?
- "this delimiter might be at fault" ?